// @workspace/api-client-react/src/generated/api.schemas.ts
/**
 * Generated by orval v8.9.1 🍺
 * Do not edit manually.
 * Api
 * NXA Growth Engine API
 * OpenAPI spec version: 0.1.0
 */
export interface HealthStatus {
  status: string;
}

export interface PlacesSearchInput {
  query: string;
  /** @nullable */
  city?: string | null;
  /** @nullable */
  state?: string | null;
  /**
     * Radius in meters
     * @nullable
     */
  radius?: number | null;
  /** @nullable */
  minRating?: number | null;
  /** @nullable */
  hasWebsite?: boolean | null;
}

export interface PlaceResult {
  placeId: string;
  name: string;
  address: string;
  /** @nullable */
  city?: string | null;
  /** @nullable */
  state?: string | null;
  /** @nullable */
  phone?: string | null;
  /** @nullable */
  website?: string | null;
  /** @nullable */
  rating?: number | null;
  /** @nullable */
  reviewCount?: number | null;
  category: string;
  /** @nullable */
  latitude?: number | null;
  /** @nullable */
  longitude?: number | null;
  /**
     * NXA opportunity score 0-100
     * @nullable
     */
  nxaScore?: number | null;
}

export interface SearchHistory {
  id: number;
  query: string;
  /** @nullable */
  city?: string | null;
  /** @nullable */
  state?: string | null;
  resultCount: number;
  createdAt: string;
}

export type LeadStatus = typeof LeadStatus[keyof typeof LeadStatus];


export const LeadStatus = {
  new: 'new',
  contacted: 'contacted',
  negotiating: 'negotiating',
  proposal: 'proposal',
  closed: 'closed',
  lost: 'lost',
} as const;

export interface Lead {
  id: number;
  /** @nullable */
  placeId?: string | null;
  name: string;
  /** @nullable */
  address?: string | null;
  /** @nullable */
  city?: string | null;
  /** @nullable */
  state?: string | null;
  /** @nullable */
  phone?: string | null;
  /** @nullable */
  whatsapp?: string | null;
  /** @nullable */
  instagram?: string | null;
  /** @nullable */
  website?: string | null;
  category: string;
  /** @nullable */
  segment?: string | null;
  /** @nullable */
  rating?: number | null;
  /** @nullable */
  reviewCount?: number | null;
  nxaScore: number;
  status: LeadStatus;
  /** @nullable */
  opportunityScore?: number | null;
  /** @nullable */
  automationFitScore?: number | null;
  /** @nullable */
  digitalMaturityScore?: number | null;
  /** @nullable */
  diagnosis?: string | null;
  /** @nullable */
  nextBestAction?: string | null;
  /** @nullable */
  latitude?: number | null;
  /** @nullable */
  longitude?: number | null;
  createdAt: string;
  updatedAt?: string;
}

export interface LeadInput {
  /** @nullable */
  placeId?: string | null;
  name: string;
  /** @nullable */
  address?: string | null;
  /** @nullable */
  city?: string | null;
  /** @nullable */
  state?: string | null;
  /** @nullable */
  phone?: string | null;
  /** @nullable */
  whatsapp?: string | null;
  /** @nullable */
  instagram?: string | null;
  /** @nullable */
  website?: string | null;
  category: string;
  /** @nullable */
  segment?: string | null;
  /** @nullable */
  rating?: number | null;
  /** @nullable */
  reviewCount?: number | null;
  nxaScore?: number;
  /** @nullable */
  latitude?: number | null;
  /** @nullable */
  longitude?: number | null;
}

/**
 * @nullable
 */
export type LeadUpdateStatus = typeof LeadUpdateStatus[keyof typeof LeadUpdateStatus] | null;


export const LeadUpdateStatus = {
  new: 'new',
  contacted: 'contacted',
  negotiating: 'negotiating',
  proposal: 'proposal',
  closed: 'closed',
  lost: 'lost',
} as const;

export interface LeadUpdate {
  /** @nullable */
  name?: string | null;
  /** @nullable */
  phone?: string | null;
  /** @nullable */
  whatsapp?: string | null;
  /** @nullable */
  instagram?: string | null;
  /** @nullable */
  website?: string | null;
  /** @nullable */
  segment?: string | null;
  /** @nullable */
  status?: LeadUpdateStatus;
  /** @nullable */
  nxaScore?: number | null;
}

export interface LeadAnalysis {
  opportunityScore: number;
  automationFitScore: number;
  digitalMaturityScore: number;
  diagnosis: string;
  nextBestAction: string;
}

export interface RadarLead {
  id: number;
  name: string;
  nxaScore: number;
  category: string;
  status: string;
  /** @nullable */
  city?: string | null;
  /** @nullable */
  latitude?: number | null;
  /** @nullable */
  longitude?: number | null;
}

export type CrmCardStage = typeof CrmCardStage[keyof typeof CrmCardStage];


export const CrmCardStage = {
  new: 'new',
  contacted: 'contacted',
  negotiating: 'negotiating',
  proposal: 'proposal',
  closed: 'closed',
  lost: 'lost',
} as const;

export interface CrmCard {
  id: number;
  leadId: number;
  leadName?: string;
  /** @nullable */
  leadCategory?: string | null;
  /** @nullable */
  leadCity?: string | null;
  stage: CrmCardStage;
  /** @nullable */
  notes?: string | null;
  /** @nullable */
  value?: number | null;
  createdAt: string;
  updatedAt?: string;
}

export type CrmCardInputStage = typeof CrmCardInputStage[keyof typeof CrmCardInputStage];


export const CrmCardInputStage = {
  new: 'new',
  contacted: 'contacted',
  negotiating: 'negotiating',
  proposal: 'proposal',
  closed: 'closed',
  lost: 'lost',
} as const;

export interface CrmCardInput {
  leadId: number;
  stage: CrmCardInputStage;
  /** @nullable */
  notes?: string | null;
  /** @nullable */
  value?: number | null;
}

/**
 * @nullable
 */
export type CrmCardUpdateStage = typeof CrmCardUpdateStage[keyof typeof CrmCardUpdateStage] | null;


export const CrmCardUpdateStage = {
  new: 'new',
  contacted: 'contacted',
  negotiating: 'negotiating',
  proposal: 'proposal',
  closed: 'closed',
  lost: 'lost',
} as const;

export interface CrmCardUpdate {
  /** @nullable */
  stage?: CrmCardUpdateStage;
  /** @nullable */
  notes?: string | null;
  /** @nullable */
  value?: number | null;
}

export interface Wallet {
  id: number;
  balance: number;
  totalEarned: number;
  totalSpent: number;
}

export type CreditTransactionType = typeof CreditTransactionType[keyof typeof CreditTransactionType];


export const CreditTransactionType = {
  debit: 'debit',
  credit: 'credit',
} as const;

export interface CreditTransaction {
  id: number;
  amount: number;
  type: CreditTransactionType;
  description: string;
  createdAt: string;
}

export type FollowupStatus = typeof FollowupStatus[keyof typeof FollowupStatus];


export const FollowupStatus = {
  pending: 'pending',
  done: 'done',
  overdue: 'overdue',
} as const;

export interface Followup {
  id: number;
  leadId: number;
  /** @nullable */
  leadName?: string | null;
  title: string;
  /** @nullable */
  notes?: string | null;
  dueDate: string;
  status: FollowupStatus;
  createdAt: string;
}

export interface FollowupInput {
  leadId: number;
  title: string;
  /** @nullable */
  notes?: string | null;
  dueDate: string;
}

/**
 * @nullable
 */
export type FollowupUpdateStatus = typeof FollowupUpdateStatus[keyof typeof FollowupUpdateStatus] | null;


export const FollowupUpdateStatus = {
  pending: 'pending',
  done: 'done',
  overdue: 'overdue',
} as const;

export interface FollowupUpdate {
  /** @nullable */
  title?: string | null;
  /** @nullable */
  notes?: string | null;
  /** @nullable */
  dueDate?: string | null;
  /** @nullable */
  status?: FollowupUpdateStatus;
}

export interface AffiliateProfile {
  id: number;
  code: string;
  link: string;
  totalConversions: number;
  totalCommission: number;
}

export type AffiliateConversionStatus = typeof AffiliateConversionStatus[keyof typeof AffiliateConversionStatus];


export const AffiliateConversionStatus = {
  pending: 'pending',
  paid: 'paid',
} as const;

export interface AffiliateConversion {
  id: number;
  commission: number;
  status: AffiliateConversionStatus;
  createdAt: string;
}

export interface StatusCount {
  status: string;
  count: number;
}

export interface SegmentCount {
  segment: string;
  count: number;
}

export interface DashboardStats {
  totalLeads: number;
  newLeads: number;
  hotLeads: number;
  creditsBalance: number;
  crmDeals: number;
  followupsDue: number;
  conversionRate: number;
  leadsByStatus?: StatusCount[];
  leadsBySegment?: SegmentCount[];
}

export type ActivityItemType = typeof ActivityItemType[keyof typeof ActivityItemType];


export const ActivityItemType = {
  lead_added: 'lead_added',
  crm_moved: 'crm_moved',
  analysis_done: 'analysis_done',
  followup_created: 'followup_created',
  search_done: 'search_done',
} as const;

export interface ActivityItem {
  id: number;
  type: ActivityItemType;
  description: string;
  /** @nullable */
  leadId?: number | null;
  createdAt: string;
}

export type ListLeadsParams = {
/**
 * @nullable
 */
status?: string | null;
/**
 * @nullable
 */
segment?: string | null;
/**
 * @nullable
 */
city?: string | null;
/**
 * @nullable
 */
minScore?: number | null;
};
